Allow configuration of SO_TIMEOUT in AgiServer

adrien-martin opened the pull request:
DefaultAgiServer use indirectly SocketConnectionFacade which has a default timeout of 4200s.
This timeout is triggered on long AGI calls like Dial.

This follows #274.

At the moment I did not modify the default value, using _null_ values.
These _null_ values are not necessary if DefaultAgiServer has a default timeout value.
Or if MAX_SOCKET_READ_TIMEOUT_MILLIS is only used by ServerSocketFacade/DefaultAgiServer (with SocketConnectionFacadeImpl(Socket)) maybe it could be moved there.

adrien-martin commented:
Hello and thanks for your review,


> Also I'm not keen on having a default of null.
> Is there any reason why we don't set the timeout default to MAX_SOCKET_READ_TIMEOUT_MILLIS

It was set to allow 0 as value, java.net.Socket#setSoTimeout(int timout) allow 0 as infinite timout.
An other value (maybe -1) could be used.
I changed the pull-request to remove null, and set default timeout to MAX_SOCKET_READ_TIMEOUT_MILLIS.

> I'm concerned about the change:
> SocketConnectionFacadeImpl(Socket socket) throws IOException
> {
> socket.setSoTimeout(MAX_SOCKET_READ_TIMEOUT_MILLIS);
> -- you removed this line --> this(socket, MAX_SOCKET_READ_TIMEOUT_MILLIS);
> }
> 
> You ctor isn't calling initialise(socket..)
> 
> Is there a reason for this?

I'm not sure to understand you here.
I added SocketConnectionFacadeImpl(Socket socket, Integer timeout) which is nearly the same as SocketConnectionFacadeImpl(Socket socket).
And made SocketConnectionFacadeImpl(Socket socket) call the other constructor.

bsutton commented:
Ignore my comment on your ctor, I read the patch backwards :<

We have had an internal discussion and would like to:
Change
MAX_SOCKET_READ_TIMEOUT_MILLIS to 3hrs.

Change
`public class ServerSocketFacadeImpl implements ServerSocketFacade
{
    private Integer socketReadTimeout = null;`
to
`public class ServerSocketFacadeImpl implements ServerSocketFacade
{
    private int socketReadTimeout = MAX_SOCKET_READ_TIMEOUT_MILLIS;`



We think 70 minutes is too short but infinite is likely to cause thread leaks if asterisk is doing something nasty.
Not many calls go for more than 3 hours so we feel this is a better balance.

adrien-martin commented:
> We have had an internal discussion and would like to:
> Change
> MAX_SOCKET_READ_TIMEOUT_MILLIS to 3hrs.

Added a commit for this.

> Change
> `public class ServerSocketFacadeImpl implements ServerSocketFacade { private Integer socketReadTimeout = null;`
> to
> `public class ServerSocketFacadeImpl implements ServerSocketFacade { private int socketReadTimeout = MAX_SOCKET_READ_TIMEOUT_MILLIS;`
> 
> We think 70 minutes is too short but infinite is likely to cause thread leaks if asterisk is doing something nasty.
> Not many calls go for more than 3 hours so we feel this is a better balance.

MAX_SOCKET_READ_TIMEOUT_MILLIS seems to be private in SocketConnectionFacadeImpl, do you want :
* MAX_SOCKET_READ_TIMEOUT_MILLIS to become public,
* to copy/paste MAX_SOCKET_READ_TIMEOUT_MILLIS in ServerSocketFacadeImpl,
* to be replaced by 0 to use the default in SocketConnectionFacadeImpl (MAX_SOCKET_READ_TIMEOUT_MILLIS).

I added 7e1d97f4 yesterday, this can be changed.

As a side note do you want all the changes in one commit or keep all the iterations ?

bsutton commented:
Make MAX_SOCKET_READ_TIMEOUT_MILLIS public.

then 
socketReadTimeout = MAX_SOCKET_READ_TIMEOUT_MILLIS;

a single commit would be nice if its not too much effort.

Thanks
Brett

adrien-martin commented:
MAX_SOCKET_READ_TIMEOUT_MILLIS was made public.

All commits reworked into one and the branch rebased on master.

bsutton commented on src/main/java/org/asteriskjava/fastagi/DefaultAgiServer.java:
```diff
@@ -62,6 +62,15 @@
+     *
+     * @see Socket#setSoTimeout(int)
+     */
+    private Integer socketReadTimeout = null;
```
```suggestion
    private int socketReadTimeout = SocketConnectionFacadeImpl.MAX_SOCKET_READ_TIMEOUT_MILLIS;
```

bsutton commented on src/main/java/org/asteriskjava/fastagi/AgiServer.java:
```diff
@@ -73,4 +73,14 @@
+     * @see java.net.Socket#setSoTimeout(int)
+     * @since 3.0.0
+     */
+    public void setSocketReadTimeout(Integer socketReadTimeout);
```
```suggestion
    public void setSocketReadTimeout(int socketReadTimeout);
```

bsutton commented on src/main/java/org/asteriskjava/fastagi/DefaultAgiServer.java:
```diff
@@ -295,9 +304,16 @@ private void loadConfig()
         }
     }
 
+    public void setSocketReadTimeout(Integer socketReadTimeout)
```
```suggestion
    public void setSocketReadTimeout(int socketReadTimeout)
```

bsutton commented on src/main/java/org/asteriskjava/util/ServerSocketFacade.java:
```diff
@@ -43,4 +43,14 @@
+     * @see java.net.Socket#setSoTimeout(int)
+     * @since 3.0.0
+     */
+    public void setSocketReadTimeout(Integer socketReadTimeout);
```
```suggestion
    public void setSocketReadTimeout(int socketReadTimeout);
```

bsutton commented on src/main/java/org/asteriskjava/util/internal/ServerSocketFacadeImpl.java:
```diff
@@ -28,14 +28,16 @@
 {
     private ServerSocket serverSocket;
 
+    private Integer socketReadTimeout = SocketConnectionFacadeImpl.MAX_SOCKET_READ_TIMEOUT_MILLIS;
```
```suggestion
    private int socketReadTimeout = SocketConnectionFacadeImpl.MAX_SOCKET_READ_TIMEOUT_MILLIS;
```

bsutton commented on src/main/java/org/asteriskjava/util/internal/ServerSocketFacadeImpl.java:
```diff
@@ -48,11 +50,16 @@ public SocketConnectionFacade accept() throws IOException
         serverSocket.close();
     }
+
+    public void setSocketReadTimeout(Integer socketReadTimeout)
```
```suggestion
    public void setSocketReadTimeout(int socketReadTimeout)
```

bsutton commented on src/test/java/org/asteriskjava/fastagi/DefaultAgiServerTest.java:
```diff
@@ -123,6 +123,10 @@ public void close() throws IOException
             closeCalls++;
         }
+
+        public void setSocketReadTimeout(Integer socketReadTimeout)
```
```suggestion
        public void setSocketReadTimeout(int socketReadTimeout)
```

adrien-martin commented:
> I've suggested that we change all Integer to int to remove any possibility of an npe.

Ah that's right sorry, I used an Integer for the null default but this is no longer the case.
Changed the commit.

bsutton commented:
Thanks for the work and your support !

adrien-martin commented:
Thanks for your time reviewing this!
